plugins/streaming/impls/jms/src/main/java/org/pentaho/di/trans/step/jms/context/ActiveMQProvider.java

```diff
@@ -60,57 +68,82 @@
   protected String buildUrl( JmsDelegate delegate ) {
     StringBuilder finalUrl = new StringBuilder( delegate.amqUrl.trim() );
 
+    if ( log.isDebug() ) {
+      log.logDebug( "ActiveMQ URL provided: " + finalUrl.toString() );
+    }
+
     // verify user hit the checkbox on the dialogue *and* also has not specified these values on the URL already
     // end result: default to SSL settings in the URL if present, otherwise use data from the security tab
     if ( delegate.sslEnabled && !finalUrl.toString().contains( "sslEnabled" ) ) {
 
-      // could already be other params on the URL
-      if ( !finalUrl.toString().contains( "?" ) ) {
-        finalUrl.append( "?" );
-      } else if ( !finalUrl.toString().endsWith( ";" ) ) {
-        finalUrl.append( ";" );
-      }
-
-      finalUrl.append( "sslEnabled=true" );
-
-      // the keystore is optional; use if client authentication is desired
-      if ( !Strings.isNullOrEmpty( delegate.sslKeystorePath ) ) {
-        finalUrl.append( ";keyStorePath=" ).append( delegate.sslKeystorePath.trim() );
-        // keystore password is always required if keystore present
-        finalUrl.append( ";keyStorePassword=" ).append( delegate.sslKeystorePassword.trim() );
-      }
-
-      // truststore always required for a client
-      finalUrl.append( ";trustStorePath=" ).append( delegate.sslTruststorePath.trim() );
-      // truststore password not required but might be present
-      if ( !Strings.isNullOrEmpty( delegate.sslTruststorePassword ) ) {
-        finalUrl.append( ";trustStorePassword=" ).append( delegate.sslTruststorePassword.trim() );
+      StringBuilder urlQuery = new StringBuilder();
+      try {
+        // could already be other params on the URL
+        if ( !finalUrl.toString().contains( "?" ) ) {
+          urlQuery.append( "?" );
+        } else if ( !finalUrl.toString().endsWith( "&" ) ) {
+          urlQuery.append( "&" );
+        }
+
+        urlQuery.append( "sslEnabled=true" );
+
+        // the keystore is optional; use if client authentication is desired
+        if ( !Strings.isNullOrEmpty( delegate.sslKeystorePath ) ) {
+          urlQuery.append( "&keyStorePath=" ).append(
+            URLEncoder.encode( delegate.sslKeystorePath.trim(), charEncoding ) );
+          // keystore password is always required if keystore present
+          urlQuery.append( "&keyStorePassword=" ).append(
+            URLEncoder.encode( delegate.sslKeystorePassword.trim(), charEncoding ) );
+        }
+
+        // truststore always required for a client
+        urlQuery.append( "&trustStorePath=" ).append(
+          URLEncoder.encode( delegate.sslTruststorePath.trim(), charEncoding ) );
+        // truststore password not required but might be present
+        if ( !Strings.isNullOrEmpty( delegate.sslTruststorePassword ) ) {
```

Comment:
Stylistic suggestion-- feel free to ignore it if you disagree.  Whenever I find myself repeating a block of code, no matter how trivial, I feel compelled to extract it out.  In this case, if the encoding method changes, or whatever, it's a single place to fix it for all.